### ISSUE TITLE
Offset when loading batch of images

### ISSUE THREAD

**Issue by theoh-io** [ultralytics/ultralytics#5066]
### Search before asking

- [X] I have searched the YOLOv8 [issues](https://github.com/ultralytics/ultralytics/issues) and found no similar bug report.

### YOLOv8 Component

Train

### Bug

I got a really strange bug where my images are offset when loaded in yolo train. By offset I mean that I translate the original image and mix it with the end of the previous one and the beginning of the next one. This bug suddenly appeared on my setup with no apparent reason. Also it seems that for the same data just changing the number of epochs to 10 randomly makes the problem disappear. here are the scrrenshot of the runs/train_batch0.jpg for the same training data launched with this command just putting epochs =10 result in no offset.
epochs to 10
![train_batch0](https://github.com/ultralytics/ultralytics/assets/50209774/12aa8f4d-4035-4ec4-8670-145fc8e0f080)

associated parameters:  task=detect, mode=train, model=yolov8l.yaml, data=cfg.yaml, epochs=10, patience=50, batch=4, imgsz=640, save=True, save_period=-1, cache=False, device=None, workers=8, project=None, name=None, exist_ok=False, pretrained=True, optimizer=auto, verbose=True, seed=0, deterministic=True, single_cls=False, rect=False, cos_lr=False, close_mosaic=10, resume=False, amp=True, fraction=1.0, profile=False, freeze=None, overlap_mask=True, mask_ratio=4, dropout=0.0, val=True, split=val, save_json=False, save_hybrid=False, conf=None, iou=0.7, max_det=300, half=False, dnn=False, plots=True, source=None, show=False, save_txt=False, save_conf=False, save_crop=False, show_labels=True, show_conf=True, vid_stride=1, stream_buffer=False, line_width=None, visualize=False, augment=False, agnostic_nms=False, classes=None, retina_masks=False, boxes=True, format=torchscript, keras=False, optimize=False, int8=False, dynamic=False, simplify=False, opset=None, workspace=4, nms=False, lr0=0.01, lrf=0.01, momentum=0.937, weight_decay=0.0005, warmup_epochs=3.0, warmup_momentum=0.8, warmup_bias_lr=0.1, box=7.5, cls=0.5, dfl=1.5, pose=12.0, kobj=1.0, label_smoothing=0.0, nbs=64, hsv_h=0.015, hsv_s=0.7, hsv_v=0.4, degrees=0.0, translate=0.1, scale=0.5, shear=0.0, perspective=0.0, flipud=0.0, fliplr=0.5, mosaic=1.0, mixup=0.0, copy_paste=0.0, cfg=None, tracker=botsort.yaml, save_dir=runs/detect/train22
Overriding model.yaml nc=80 with nc=3

any other number of epochs (150)
![train_batch0](https://github.com/ultralytics/ultralytics/assets/50209774/73b2061c-9479-40dc-8091-eec5476146ec)
associated parameters: task=detect, mode=train, model=yolov8l.yaml, data=cfg.yaml, epochs=150, patience=50, batch=4, imgsz=640, save=True, save_period=-1, cache=False, device=None, workers=8, project=None, name=None, exist_ok=False, pretrained=True, optimizer=auto, verbose=True, seed=0, deterministic=True, single_cls=False, rect=False, cos_lr=False, close_mosaic=10, resume=False, amp=True, fraction=1.0, profile=False, freeze=None, overlap_mask=True, mask_ratio=4, dropout=0.0, val=True, split=val, save_json=False, save_hybrid=False, conf=None, iou=0.7, max_det=300, half=False, dnn=False, plots=True, source=None, show=False, save_txt=False, save_conf=False, save_crop=False, show_labels=True, show_conf=True, vid_stride=1, stream_buffer=False, line_width=None, visualize=False, augment=False, agnostic_nms=False, classes=None, retina_masks=False, boxes=True, format=torchscript, keras=False, optimize=False, int8=False, dynamic=False, simplify=False, opset=None, workspace=4, nms=False, lr0=0.01, lrf=0.01, momentum=0.937, weight_decay=0.0005, warmup_epochs=3.0, warmup_momentum=0.8, warmup_bias_lr=0.1, box=7.5, cls=0.5, dfl=1.5, pose=12.0, kobj=1.0, label_smoothing=0.0, nbs=64, hsv_h=0.015, hsv_s=0.7, hsv_v=0.4, degrees=0.0, translate=0.1, scale=0.5, shear=0.0, perspective=0.0, flipud=0.0, fliplr=0.5, mosaic=1.0, mixup=0.0, copy_paste=0.0, cfg=None, tracker=botsort.yaml, save_dir=runs/detect/train23
Overriding model.yaml nc=80 with nc=3

### Environment

Ultralytics YOLOv8.0.184 🚀 Python-3.10.12 torch-2.0.1+cu117 CUDA:0 (NVIDIA GeForce RTX 4060 Laptop GPU, 8188MiB)
Setup complete ✅ (16 CPUs, 7.6 GB RAM, 185.7/1006.9 GB disk)

OS                  Linux-5.15.90.1-microsoft-standard-WSL2-x86_64-with-glibc2.35
Environment         Linux
Python              3.10.12
Install             pip
RAM                 7.61 GB
CPU                 12th Gen Intel Core(TM) i7-12650H
CUDA                11.7

matplotlib          ✅ 3.8.0>=3.3.0
numpy               ✅ 1.26.0>=1.22.2
opencv-python       ✅ 4.8.0.76>=4.6.0
pillow              ✅ 10.0.1>=7.1.2
pyyaml              ✅ 6.0.1>=5.3.1
requests            ✅ 2.31.0>=2.23.0
scipy               ✅ 1.11.2>=1.4.1
torch               ✅ 2.0.1>=1.8.0
torchvision         ✅ 0.15.2>=0.9.0
tqdm                ✅ 4.66.1>=4.64.0
pandas              ✅ 2.1.1>=1.1.4
seaborn             ✅ 0.12.2>=0.11.0
WARNING ⚠️ invalid check_version(5.9.5, ) requested, please check values.
psutil              ✅ 5.9.5
WARNING ⚠️ invalid check_version(9.0.0, ) requested, please check values.
py-cpuinfo          ✅ 9.0.0

### Additional

Maybe I missed something but I am convinced that it's an internal bug as the only parameter I changed shouldn't influence the data in any way.

### Are you willing to submit a PR?

- [X] Yes I'd like to help by submitting a PR!

**Comment by glenn-jocher** [comment]
@theoh-io this is mosaic dataloading, which I invented a few years ago and first applied to YOLOv5. It's a feature that improves training results, not a bug. Enjoy.

https://docs.ultralytics.com/datasets/detect/coco/#sample-images-and-annotations
<img width="1474" alt="Screenshot 2023-09-24 at 17 36 59" src="https://github.com/ultralytics/ultralytics/assets/26833433/8634bd09-b37c-4b55-a09f-a7039f78aef6">
